File: elixir_leetcode/lib/day10/binary_search.ex
defmodule BinarySearch do
  @spec search(nums :: [integer], target :: integer) :: integer
  def search(nums, target) do
    IO.puts('--')
    divide(nums, target, 0, length(nums))
  end


  def divide(arr, target, left, right) do
    index = floor((left + right) / 2)
    # Enum.at/3 跟 Enum.fetch/2 相比，可以給 default 值，那他媽的 Enum.fetch/2 要幹嘛？？？
    num = Enum.at(arr, index, nil)

    cond do
      # 會拿不到值，這時候就 return -1
      num == nil -> -1
      left >= right -> -1
      num === target -> index
      target > num ->
        divide(arr, target, index + 1, right)
      true ->
        divide(arr, target, left, index)
    end
  end


end

File: ts_leetcode/day10/binary_search.ts
export function search(nums: number[], target: number): number {
  let l: number = 0;
  let r: number = nums.length;

  do {
    const index = Math.floor((r + l) / 2);
    if (nums[index] === target) return index;
    if (target > nums[index]) {
      l = index + 1;
    } else {
      r = index;
    }
  } while (l < r);

  return -1;
}
